rapid7/metasploit-payloads
# OpenJDK reflection fix

This PR aims to resolve #647 - allowing the Java Meterpreter to run on the latest OpenJDK. 

Contract-breaking reflection (i.e. forcibly changing private fields to be public, then calling them) was prevented in this JDK; this was previously used to hijack the URL Class Loading process to have "URLs" to in-memory resources. 

The new approach avoids the use of `URLClassLoaders` entirely, since (being unable to use the standard global `URLStreamHandlerFactory` for risk of conflicting with whatever application we're injecting into), we seem to need to do some hackery to have the `URL` class handle an "in-memory URL protocol". Without the use of invasive reflection, I found it much easier to do away with that entirely, and just have our own `ClassLoader` type. We get more control that way.

Testing/landing this work requires a corresponding PR in the metasploit-framework repo, to just send the correct class file(s) to the stager. You'll need to checkout the appropriate branch, and `make` from this repo's directory to set up the test situation. That PR is here: https://github.com/rapid7/metasploit-framework/pull/18355

I've tested it back to Java 1.5, which worked (although the HTTPS payload failed; Wireshark indicated it was probably a SSL versioning error; maybe new ruby doesn't support the old java ciphers?).

The README in this repo says that this payload should work on JRE 1.2, but the existing payload files didn't seem to work on that Java version(`Exception in thread "main" java.lang.UnsupportedClassVersionError: metasploit/Payload (Unsupported major.minor version 49.0)`). The build process only seems to check back to 1.5; so I'm assuming the actual back-compat requirement is 1.5.

Test cases:
- Run payloads on the latest OpenJDK
- Verify it still works on old JDK instances.
- Verify each of the payload types result in a usable shell (bind and reverse; shell and meterp; and the one staged payload: bind shell)
- Verify interactivity works with each of them.
